src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -169,19 +177,62 @@ internal string CreateCacheKey(CachedVaryBy varyBy)
                             value = "null";
                         }
 
-                        builder.Append(";")
+                        builder.Append(KeyDelimiter)
                             .Append(header)
                             .Append("=")
                             .Append(value);
                     }
                 }
-                // TODO: Parse querystring params
+
+                // Vary by query params
+                if (varyBy?.Params.Count > 0)
+                {
+                    // Append a value denoting the start of the query parameter segment of the cache key
+                    builder.Append(KeyDelimiter)
+                        .Append('Q');
+
+                    if (varyBy.Params.Count == 1 && string.Equals(varyBy.Params[0], "*"))
+                    {
+                        // Vary by all available query params
+                        foreach (var query in _httpContext.Request.Query.OrderBy(q => q.Key))
+                        {
+                            builder.Append(KeyDelimiter)
+                                .Append(query.Key.ToLower())
+                                .Append("=")
+                                .Append(query.Value);
```

Comment:
It's ambiguous what we need to do here, we'll discuss offline in a design meeting.